src/slider.tsx

```diff
@@ -1,164 +1,428 @@
-                ? slider.left - 10 - newValue * multiplyValue
-                : 10 - newValue * 0.2;
-            return `calc(${newValue}% + (${newPosition}px))`;
+type SliderProps = BaseSliderProps & ExclusifyUnion<SliderWithStepProps | SliderWithValuesProps>;
+
+const getSliderValueAsPercentage = (value: number, min: number, max: number) => {
+    return min >= max ? 0 : (value - min) / (max - min);
+};
+
+const getValueInRange = (isPercentage: boolean, min: number, max: number, step: number, value?: number) => {
+    const getRealValue = (value?: number) => {
+        if (value === undefined) {
+            return min;
+        }
+        const realValue = isPercentage ? min + (max - min) * value : value;
+        return Math.max(min, Math.min(max, realValue));
+    };
+
+    if (min >= max) {
+        return min;
+    }
+
+    const currentValue = getRealValue(value);
+    const valueRoundedDown = min + Math.floor((currentValue - min) / step) * step;
+    const valueRoundedUp = min + Math.ceil((currentValue - min) / step) * step;
+
+    return valueRoundedUp <= max && valueRoundedUp - currentValue <= currentValue - valueRoundedDown
+        ? valueRoundedUp
+        : valueRoundedDown;
+};
+
+const getClosestValidValue = (value: number, values?: Array<number>) => {
+    if (!values) {
+        return value;
+    }
+
+    let closestIndex = 0;
+    values.forEach((currentValue, index) => {
+        if (values && Math.abs(currentValue - value) <= Math.abs(values[closestIndex] - value)) {
+            closestIndex = index;
+        }
+    });
+    return closestIndex;
+};
+
+const Slider = React.forwardRef<HTMLInputElement, SliderProps>(
+    (
+        {
+            values,
+            step = 1,
+            steps,
+            getStepArrayIndex,
+            min = 0,
+            max = 100,
+            'aria-label': ariaLabel,
+            'aria-labelledby': ariaLabelledBy,
+            id,
+            dataAttributes,
+            tooltip,
+            ...props
         },
-        [valueRanger, minSlider, maxSlider]
-    );
-
-    const getApproximation = React.useCallback(
-        (value: number) => {
-            let finalValue = value;
-            if (Array.isArray(steps)) {
-                finalValue = steps.reduce((a, b) => {
-                    return Math.abs(b - value) < Math.abs(a - value) ? b : a;
-                });
+        ref
+    ) => {
+        if (values) {
+            if (values.length === 0) {
+                values = undefined;
+            } else {
+                max = values.length - 1;
             }
-            return finalValue;
-        },
-        [steps]
-    );
-
-    const getOrder = React.useCallback(() => {
-        if (Array.isArray(steps)) {
-            steps.sort((a, b) => {
-                return a - b;
-            });
         }
-    }, [steps]);
 
-    const handleSlider = (value: number) => {
-        setValueRanger(value);
-    };
+        step = step <= 0 ? 1 : step;
```

Comment:
consider this:

```js
step = step | 0; // to avoid float or non numeric steps
step = step <= 0 ? 1 : step; // your current check
step = step > max ? max : step; // also avoid too large steps, which would make the final value unreachable
```

Reply:
Discussed offline.

I will add the second condition, but the third one is unnecessary. As of now, it behaves exactly like the native html input, and I don't think we should change that. Also, even if we used it, there are still cases when you can't reach the maximum value (min=0, max=10, step=3).